---

# you may want to add your name here and document title
output:
  html_document: default
  pdf_document: default
---
**Part 1: 

**Data inspection**

#Load packages
```{r}
library("tidyverse")
library("dplyr")
library(data.table)
library(ggplot2)
```

#Data Preparation
#Read the data files (fang.mf and snp.mf) we are going to use.Because a header is present we will set the “header” argument to TRUE so that R knows the first line contains a header.

```{r}
fang.mf <- read.delim("C:/Users/mfakude/OneDrive - Iowa State University/Documents/EEOB546_R_lesson/fang_et_al_genotypes.txt", header = TRUE)
snp.mf <- read.delim ("C:/Users/mfakude/OneDrive - Iowa State University/Documents/EEOB546_R_lesson/snp_position.txt", header = TRUE)
```
#Data inspection of fang_et_al_genotypes.txt (Now named fang.mf)
```{r}
head(fang.mf) #shows first 6 rows
tail(fang.mf) #shows last 6 rows
dim(fang.mf) #returns the dimensions of data frame (number of rows and number of columns)
nrow(fang.mf) #number of rows # Is this not a repetition of `dim`? 
ncol(fang.mf) #number of columns
str(fang.mf) #structure of data frame 
names(fang.mf) #shows the names attribute for a data frame, which gives the column names
sapply(fang.mf, class) #shows the class of each column in the data frame
ls(fang.mf)#shows objects in the objects in the frame
```
##Data inspection of snp_position.txt (Now named snp.mf)
```{r}
head(snp.mf) #shows first 6 rows
tail(snp.mf) #shows last 6 rows
dim(snp.mf) #returns the dimensions of data frame (number of rows and number of columns)
nrow(snp.mf) #number of rows
ncol(snp.mf) #number of columns
str(snp.mf) #structure of data frame - name, type and preview of data in each column
names(snp.mf) #shows the names attribute for a data frame, which gives the column names
sapply(snp.mf, class) #shows the class of each column in the data frame
ls(fang.mf) #shows objects in the objects in the frame
```

**Data Processing** 

#Extract 3 columns (SNP_ID, Chromosome, Position) from snp.mf file and create a new file named snp_extracted.
#Save the newly created file containg the extracted columns and view it in a spreadsheet-style.
#Extract maize genotypes (ZMMIL, ZMMLR, ZMMMR) on 'Group' column from fang.mf file and create a new file named 'maize'.
#Save the newly created file containg the extracted maize and view it.
```{r}
snp_extracted <- snp.mf[,c("SNP_ID", "Chromosome", "Position")]
write.table(snp_extracted,"./Data_processing/snp_extracted.txt")
View(snp_extracted)
maize <- fang.mf %>% filter(Group == "ZMMIL" | Group == "ZMMLR" | Group == "ZMMMR")
write.table(maize,"./Data_processing/maize.txt")
```
#Extract teosinte genotypes (ZMPBA, ZMPIL, ZMPJA) on 'Group' column from fang.mf file and create a new file named 'teosinte'.
#Save the newly created file containg the extracted teosinte and view it.

```{r}
teosinte <- fang.mf %>% filter(Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
write.table(teosinte,"./Data_processing/teosinte.txt")
```
##Join object maize file and with snp_extracted file (For Maize)
#Transpose maize file
#Save the transposed maize file
#Name the first row SNP_ID
#Join object maize file and with snp_extracted file
#Save the joined maize file and view file
```{r}
maize_transposed <- t(maize[,-c(1:3)]) %>% as.data.frame()
write.table(maize_transposed,"./Data_processing/maize_transposed.txt")
maize_transposed$SNP_ID <- rownames(maize_transposed)
maize_final <- merge(snp_extracted,maize_transposed,by = 'SNP_ID')
colnames(maize_final)[4:ncol(maize_final)] <- as.character(maize$Sample_ID)
maize_final[,-c(1:3)] <- lapply(maize_final[,-c(1:3)],as.character)
unique(maize_final$Chromosome)
write.table(maize_final,"./Data_processing/maize_final.txt")
```
#Join object teosinte file and with snp_extracted file (For teosinte)
#Transpose teosinte file
#Save the transposed maize file
#Name the first row SNP_ID
#Join object teosinte file and with snp_extracted file
#Save the joined teosinte file
```{r}
teosinte_transposed <- t(teosinte[,-c(1:3)]) %>% as.data.frame()
write.table(teosinte,"./Data_processing/teosinte_transposed.txt")
teosinte_transposed$SNP_ID <- rownames(teosinte_transposed)
teosinte_final <- merge(snp_extracted,teosinte_transposed,by = 'SNP_ID')
colnames(teosinte_final)[4:ncol(teosinte_final)]<-as.character(teosinte$Sample_I)
write.table(teosinte_final,"./Data_processing/teosinte_final.txt")
```

##Subseting maize/teosinte chromosomes with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

##files with chromosomes with SNPs arranged on increasing position values and with missing data encoded by this symbol '?' are name with an Q

##Files with chromosomes with SNPs arranged on decreasing position values and with missing data encoded by this symbol '-' are name with an H. For example, maize_chr1_Q or maize_chr1_H/teosinte_chr1_Q or teosinte_chr1_H

#subset by chromosome and arrange in ascending order
#replace symbol ('?/?') with symbol ('?')
#save files and name each file by chromosome
#subset by chromosome and arrange in descending order
#replace symbol ('?/?') with symbol('?')
#save files and name each file by chromosome number and symbol
#for (chr in 1:10) 
```{r}

for (chr in 1:10){
  maize_chr <- subset(maize_final,Chromosome == chr) %>% arrange(Position) 
  maize_chr[maize_chr == '?/?'] <- '?' 
  write.table(maize_chr,file=paste("./Maize_chromosome/maize_chr",chr,"_Q.txt",sep = "")) 
  maize_chr_d <- subset(maize_final,Chromosome == chr) %>% arrange(desc(Position))
  maize_chr_d[maize_chr_d == '?/?'] <- '-' 
  write.table(maize_chr_d,file = paste("./Maize_chromosome/maize_chr",chr,"_H.txt",sep = ""))
  
  teosinte_chr <- subset(teosinte_final,Chromosome == chr) %>% arrange(Position)
  teosinte_chr[teosinte_chr == '?/?'] <- '?'
  write.table(teosinte_chr,file = paste("./Teosinte_chromosome/teosinte_chr",chr,"_Q.txt",sep = ""))
  teosinte_chr_d <- subset(teosinte_final,Chromosome == chr) %>% arrange(desc(Position))
  teosinte_chr_d[teosinte_chr_d == '?/?'] <- '-'
  write.table(teosinte_chr_d,file = paste("./Teosinte_chromosome/teosinte_chr",chr,"_H.txt",sep = ""))
}
```

**Part 2**
  
**Data visualization**
  

##1.(i) Plot SNPs per chromososome
#Filter the original snp_position file by position and command to filter for items less than infinity
#Pipe the filtering to produce a bar graph of total SNP counts per chromosome
#Label the X-axis and the y-axis and save the plot
  
```{r}  
library(ggplot2)
snp.mf %>% 
  filter(Position < 10e1000) %>% 
  ggplot(aes(as.double(Chromosome))) +
  geom_bar(fill = 'purple', color = 'yellow') + 
  geom_text(stat = 'count', aes(label = ..count..), vjust = -1) +
  scale_x_continuous(breaks = 1:10) +
  theme_replace() +
  ggtitle("snp count vs Chromosome") +
  ylab('Number of snp') +
  xlab('Chromosome') 
ggsave("./Data_visualization/Number of snp_chromosome.png")
```
#1(ii)Plotting SNP distribution across the chromosome
#Filter the position for anything less than infinity and map them to position per Mb
#Pipe the latter to produce a histogram that show the distributionof SNPs in a chromosome  
#Color the histogram with a purple color
#Plot each chromosome using facet wrap
#Label the X-axis and the y-axis and save the plot
```{r}
snp.mf %>% filter(Position < 10e1000) %>% 
  ggplot(aes(as.double(Position)/1000000)) +
  geom_histogram(aes(y = ..density..), color = 'purple', fill = "purple", alpha = 0.4, bins = 20) + 
  geom_density(aes(as.double(Position)/1000000), color = "yellow") + 
  facet_wrap(~ as.double(Chromosome), scales = "free") +
  theme_replace() +
  ggtitle("SNP distribution across chromosomes") +
  xlab('Position (Mb)') +
  ylab('Density of SNPs')
ggsave("./Data_visualization/SNP_distribution_across_chromosomes.png")
```
##2 Missing data and amount of heterozygosity

##2(i) Proportion of homozygous and heterozygous by Sample ID 
#Remove the 2nd tha the 3rd column in fang.mf and pipe it pivot_long by Sample_ID, to produce a tibble of each SNP and its genotypic value 
#Mutate the file to add a colum with homozygous and heterozygous SNPs 

#Plot graph (SNPs) by Sample ID and save it
```{r}
mutate_Sampleid <- 
  fang.mf %>% select(-JG_OTU, -Group) %>%   
  pivot_longer(!Sample_ID) %>% 
  mutate(Allele = ifelse(value %in% c('C/C', 'G/G', 'A/A', 'T/T'), 'Homozygote', ifelse(value == '?/?', 'MD', 'Heterozygote')))
View(mutate_Sampleid)

library(RColorBrewer)
color_plots <- brewer.pal(3, "Set3")
mutate_Sampleid %>% group_by(Sample_ID) %>%  count(Allele) %>% 
  ggplot(aes(fill = Allele, y = n, x = Sample_ID)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = color_plots) +
  ggtitle("Ratio of homozygotes, heterozygotes and missing data by genotype ") +
  ylab('Ratio') +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
ggsave("./Data_visualization/Gene_ratio by sample.png")
```

##2(ii) Ratio of homozygous,heterozygotes and missing data by Group
#Remove the 1st and the 3rd column in fang.mf file and pipe it to pivot_long by Group, to create a tibble of each group and its genotypic value 
#Mutate the file to add a column with homozygous and hetrozygous SNPs 

#Plot graph (SNPs) by group and save it 
```{r}
mutate_Group <- 
  fang.mf %>% select(-JG_OTU, -Sample_ID) %>%   
  pivot_longer(!Group) %>% 
  mutate(Allele = ifelse(value %in% c('C/C', 'G/G', 'A/A', 'T/T'), 'Homozygote', ifelse(value == '?/?', 'MD', 'Heterozygote')))  
View(mutate_Group)

mutate_Group %>% group_by(Group) %>%  count(Allele) %>% 
  ggplot(aes(fill = Allele, y = n, x = Group)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = color_plots) +
  ggtitle("Ratio of homozygotes, heterozygotes and missing data by Group ") +
  ylab('Ratio') 
ggsave("./Data_visualization/Ratio of zygosity and missing data.png")
```
#3.My_plot_ Plot graph of proprotion of homozygotes by Group and save it
# is this your own or is aprt of the requirement?
```{r}
color2 <- brewer.pal(4, "Spectral")
mutate_Group %>% filter(Allele == "Homozygote") %>% group_by(Group) %>%  count(value) %>% 
  ggplot(aes(fill = value, y = n, x = Group)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = color2) +
  ggtitle("Ratio of homozygotes in each group") +
  ylab('Ratio') +
  theme_bw()
ggsave("./Data_visualization/my_plot.png")
# Isnt the file pathway supposed to be "absolute" 
```
